### DISCUSSION TITLE
forcing exp operator to be used in special form

### DISCUSSION THREAD

**Discussion by arasugurlu** [MilesCranmer/PySR#1078] [Q&A] [answered]
I'm working with **PySR** (latest version, using Julia backend) and I noticed that
the `exp` operator is *never used* in the discovered equations, even though it is included
in the `unary_operators` list.

from pysr import PySRRegressor
import numpy as np

X = data
y = data

model = PySRRegressor(
    unary_operators=["exp"],
    binary_operators=["+", "-", "*", "/"],
    niterations=100,
    procs=4,
    populations=20,
    population_size=100,
    complexity_of_operators={"exp": 1},
)

model.fit(X, y)
print(model)

Even though the data clearly follows an exponential law,
the discovered expressions never contain exp().
It keeps returning polynomial or multiplicative combinations instead.

I have tried,

Increasing niterations, population_size, and populations

Adjusting complexity_of_operators={"exp": 1}

Using only "exp" as unary operator

At the end, even if exp shows up, it only shows up as exp(K) where K is a constant real number. I need to force "exp" operator to be used as exp (x1/x2) form where x1 and x2 the features given from the dataset.

I want my analytical equation to be in exp(x1/x2) form but i never made it. 

Thanks for this amazing package

**Comment by MilesCranmer** [comment]
Does x2 ever get very close to 0? It could just blow up and hurt convergence, maybe.

**Reply by arasugurlu** [reply]
No, it does not. The minimum value of x2 is 2.3

**Reply by MilesCranmer** [reply]
Not sure then. Very weird. Can you share your data or another reproducible example

**Reply by arasugurlu** [reply]
Sorry for the late reply. 
I am now sharing the first 20 row. the feature x1 is the 5th column, and the feature x2 is the 1st column. 
No matter what I do, PySR never uses exp. 
Also, I use dimensional constraints, because I want my analytical equation to be dimensionally consistent. Furthermore, the dimensions of x1 and x2 are the same, [m], therefore, PySR should use the 1st and the 5th column, since it will not create any dimensional inconsistency.

3.810000	214.000000	74.900000	136.400000	15.487783	13.469412	1.200000
4.730000	249.000000	115.900000	123.400000	10.343672	 8.106783	13.110000
5.290000	513.500000	143.900000	254.400000	24.149624	21.655733	1.030000
3.750000	505.000000	168.400000	163.400000	6.386264	         5.377547	21.250000
5.400000	383.000000	50.400000	125.400000	14.043067	13.105884	7.520000
3.740000	584.000000	96.900000	77.400000	5.763707	         4.885063	38.620000
2.940000	561.000000	203.900000	242.900000	4.816926        	4.125744	        9.610000
4.110000	414.000000	53.400000	80.400000	5.499353	        4.890981	       18.670000
3.120000	229.500000	59.900000	116.900000	2.707272	        2.449124 	54.840000
5.310000	418.000000	223.900000	80.400000	7.544074  	5.695700          15.870000
5.770000	487.500000	213.400000	210.400000	14.556022	11.991710	1.780000
5.300000	546.500000	190.900000	195.400000	17.952777	15.172839	2.740000
4.820000	459.000000	67.400000	38.400000	6.010049	        5.020833	        36.770000
4.940000	515.500000	46.900000	67.900000	10.880470	9.518432	        11.580000
4.040000	545.000000	89.900000	156.900000	22.394042	20.284650	0.930000
5.560000	270.000000	206.900000	85.900000	5.954621	        3.804390	        98.410000
3.140000	214.000000	55.400000	90.400000	2.293034	       2.042386	        97.100000
4.530000	372.500000	229.900000	181.400000	7.502371	        5.445340  	35.660000
6.070000	357.000000	153.900000	68.900000	10.519023	8.159822	        10.060000
4.220000	214.000000	78.900000	73.900000	3.375094	       2.803827  	74.870000

Thank you for your support

**Reply by MilesCranmer** [reply]
Sorry, the data is unformatted, please use triple backticks

**Reply by arasugurlu** [reply]
Sorry
```
3.810000 214.000000 74.900000 136.400000 15.487783 13.469412 1.200000
4.730000 249.000000 115.900000 123.400000 10.343672 8.106783 13.110000
5.290000 513.500000 143.900000 254.400000 24.149624 21.655733 1.030000
3.750000 505.000000 168.400000 163.400000 6.386264 5.377547 21.250000
5.400000 383.000000 50.400000 125.400000 14.043067 13.105884 7.520000
3.740000 584.000000 96.900000 77.400000 5.763707 4.885063 38.620000
2.940000 561.000000 203.90000 242.900000 4.816926 4.125744 9.610000
4.110000 414.000000 53.400000 80.400000 5.499353 4.890981 18.670000
3.120000 229.500000 59.900000 116.900000 2.707272 2.449124 54.840000
5.310000 418.000000 223.900000 80.400000 7.544074 5.695700 15.870000
5.770000 487.500000 213.400000 210.400000 14.556022 11.991710 1.780000
5.300000 546.500000 190.900000 195.400000 17.952777 15.172839 2.740000
4.820000 459.000000 67.400000 38.400000 6.010049 5.020833 36.770000
4.940000 515.500000 46.900000 67.900000 10.880470 9.518432 11.580000
4.040000 545.000000 89.900000 156.900000 22.394042 20.284650 0.930000
5.560000 270.000000 206.900000 85.900000 5.954621 3.804390 98.410000
3.140000 214.000000 55.400000 90.400000 2.293034 2.042386 97.100000
4.530000 372.500000 229.900000 181.400000 7.502371 5.445340 35.660000
6.070000 357.000000 153.900000 68.900000 10.519023 8.159822 10.060000
4.220000 214.000000 78.900000 73.900000 3.375094 2.803827 74.870000
```

**Reply by MilesCranmer** [reply]
Sorry, What is y? I see 7 columns

**Reply by MilesCranmer** [reply]
Also you mentioned using dimensional constraints, but I don't see any in your code. Can you please create a MRE?

https://stackoverflow.com/help/minimal-reproducible-example

Without a MRE I cannot help, sorry.

**Reply by arasugurlu** [reply]
The y value is the 7th column, it can also be understood from the line " y_val = float(parts[6]) ". 
I apologize for bothering you, I didn't know what format to use.
Here, **I try to use exp(x6) = exp(x4/x0) and exp (x7) = exp(x5/x0)** if possible.
It would be also great to optimize those expressions  with a constant number, such as, **exp(K*x4/x0) where K is a real constant number**.
The dimensions of the constants also can be  seen in the comments. I really appreciate your help. You can see the first 20 rows above.


import numpy as np
import matplotlib.pyplot as plt
from pysr import PySRRegressor

--- Example data ---
file_path = r"/arasugurlu/PySR-2/train_set.txt"


with open(file_path, "r", encoding="utf-8") as file:
    for line in file:
        parts = line.strip().split()
        if len(parts) != 7:
            continue
        try:
           
            x0 = float(parts[0])  # m
            x1 = float(parts[1])  # N/m^2
            x2 = float(parts[2])  # N/m^2
            x3 = float(parts[3])  # N/m^2
            x4 = float(parts[4])  # m
            x5 = float(parts[5])  # m
            x6 = x4 / x0          # dimensionless
            x7 = x5 / x0          # dimensionless
            x8 = x4 / x5          # dimensionless
            g = 3*x3*(x1-x2)/(x1-x2+4*x3) # N/m^2
  	    y_val = float(parts[6]) # N/m^2

	    X.append([x0, g, x6, x7])
            y.append(y_val)

X = np.array(X)
y = np.array(y)

X_units = ["m","N/m^2","",""]
y_units = "N/m^2"

--- Example elementwise loss (it is randomly chosen)---
elementwise_loss = """
function loss_fnc(predictions, targets)
    total_loss = 0.0
    for (p, t) in zip(predictions, targets)
        # simple relative error
        total_loss += abs(p - t) / (abs(t) + 1e-6)
    end
    return total_loss
end
"""

--- PySR model ---
model = PySRRegressor(
    procs=64,
    populations=192,
    population_size=192,
    ncycles_per_iteration=500,
    niterations=1000,
    maxsize=25,
    maxdepth=25,
    parsimony=0.001,
    binary_operators=["*", "+", "-", "/"],
    unary_operators=["exp"],
    dimensionless_constants_only=True,
    weight_randomize=0.2,
    weight_optimize=0.001,
    warm_start=True,
    elementwise_loss=elementwise_loss
)

Fit model
model.fit(X, y)
y_pred = model.predict(X)

**Reply by MilesCranmer** [reply]
Hi, I am still unable to reproduce it, sorry. Please read through https://stackoverflow.com/help/minimal-reproducible-example for instructions on how to create such an example. Try to make it minimal and reproducible, so if I copy-paste your code on my machine, I see the same issue. You could have the dataset written in the script itself, for example.

**Reply by MilesCranmer** [reply]
And also please use triple backticks for pasting code

**Reply by MilesCranmer** [reply, marked as answer]
Oh. I think I can figure it out by looking at your code. You have `dimensionless_constants_only=True`. Which makes your problem impossible to solve. Because `K * x4/x0` has dimensions, and therefore `exp(K * x4 / x0)` is invalid. 

Hope this helps!

**Reply by arasugurlu** [reply]
Thank you, now i can use, however, there is dimensional inconsistency. How can i solve this?
